NIFIREG-300 Added nifi-registry-revision modules containg NiFi's Revi…

bbende opened the pull request:
…sionManager concept with a JDBC implementation and supporting utility modules

This pull request doesn't change any functionality in the application, it just establishes the nifi-registry-revision modules. It is probably easiest to review them by looking through the modules in an IDE to see how the pieces fit together, starting with nifi-registry-revision-api which directly correlates to NiFi's revision API (potentially they could share this API module once released).

kevdoran commented on pom.xml:
```diff
@@ -94,13 +94,14 @@
+        <jackson.version>2.9.9</jackson.version>
+        <spring.boot.version>2.1.6.RELEASE</spring.boot.version>
+        <spring.security.version>5.1.5.RELEASE</spring.security.version>
+        <flyway.version>5.2.4</flyway.version>
```
Do any of these dep version changes require NOTICE file changes? not sure if we mention version numbers there or not.

bbende replied:
I will double check, but they only would require changes if they brought in need dependencies that also required mentions in the NOTICE, or I guess if the copyright years changed.

kevdoran commented on nifi-registry-core/nifi-registry-revision/nifi-registry-revision-spring-jdbc/pom.xml:
```diff
@@ -0,0 +1,62 @@
+            <groupId>org.springframework.boot</groupId>
+            <artifactId>spring-boot-starter-jdbc</artifactId>
+            <version>${spring.boot.version}</version>
+        </dependency>
```
Is `spring-boot-starter-jdbc`necessary? Unless I'm missing something it seems like only `spring-jdbc` is needed.

bbende replied:
Good point, I'll try using just spring-jdbc and see what happens.

The only part that makes me hesitant is then we basically have to maintain a version of spring-jdbc, independently of the version of spring-boot.

bbende replied:
Seems to work fine just using spring-jdbc so I'll leave it with that.

kevdoran commented on nifi-registry-core/nifi-registry-revision/nifi-registry-revision-entity-model/src/main/java/org/apache/nifi/registry/revision/entity/RevisionInfo.java:
```diff
@@ -0,0 +1,79 @@
+import io.swagger.annotations.ApiModelProperty;
+
+@ApiModel(description = "The revision information for an entity managed through the REST API.")
+public class RevisionInfo {
```
Disregard this comment. I see the classes are different now.

bbende commented on nifi-registry-core/nifi-registry-revision/nifi-registry-revision-api/src/main/java/org/apache/nifi/registry/revision/api/Revision.java:
```diff
@@ -0,0 +1,111 @@
+        if (this.version != null && this.version.equals(thatRevision.version)) {
+            return true;
+        } else {
+            return clientId != null && !clientId.trim().isEmpty() && clientId.equals(thatRevision.getClientId());
```
I think you are right that the trim isn't helpful here, this was existing NiFi code that was ported so I will probably leave it as is until we know of an issue that requires changing it

bbende commented on nifi-registry-core/nifi-registry-revision/nifi-registry-revision-common/src/main/java/org/apache/nifi/registry/revision/standard/RevisionComparator.java:
```diff
@@ -0,0 +1,40 @@
+            return entityComparison;
+        }
+
+        final int clientComparison = o1.getClientId().compareTo(o2.getClientId());
```
Good catch, I think the reason this ends up never happening is because in the "clientId" parameter class in the REST layer, we have:

```
public ClientIdParameter(String clientId) {
        if (clientId == null || clientId.trim().isEmpty()) {
            this.clientId = UUID.randomUUID().toString();
        } else {
            this.clientId = clientId;
        }
    }
```
So it always ends up non-null, but I'm going to update this code with something like this:
```

final Comparator<String> nullSafeStringComparator = 
        Comparator.nullsFirst(String::compareTo);

final int clientComparison = Objects.compare(
        o1.getClientId(), o2.getClientId(), nullSafeStringComparator);
```

bbende commented:
@kevdoran pushed a new commit that addresses the comments (comparator null check, NOTICE updates, use of spring-jdbc). 

I also added a note in the javadoc of all the API classes stating this is really a framework level API and not a public extension point, so we are free to evolve as necessary.

Let me know if there is anything else.
